diff --git a/dwmblocks.c b/dwmblocks.c
index befd252..0f13f01 100644
--- a/dwmblocks.c
+++ b/dwmblocks.c
@@ -53,7 +53,6 @@ static void (*writestatus) () = pstdout;
 static char statusbar[LENGTH(blocks)][CMDLENGTH] = {0};
 static char statusstr[2][STATUSLENGTH];
 static int statusContinue = 1;
-static int returnStatus = 0;
 
 //opens process *cmd and stores output in *output
 void getcmd(const Block *block, char *output)
