diff --git a/config.def.h b/config.def.h
index defa426..d823ab6 100644
--- a/config.def.h
+++ b/config.def.h
@@ -15,6 +15,8 @@ static const int  tabwidth      = 200;
 static const Bool foreground    = True;
 static       Bool urgentswitch  = False;
 
+static const int  barHeight     = 24;
+
 /*
  * Where to place a new tab when it is opened. When npisrelative is True,
  * then the current position is changed + newposition. If npisrelative
diff --git a/tabbed.c b/tabbed.c
index eafe28a..c031630 100644
--- a/tabbed.c
+++ b/tabbed.c
@@ -984,7 +984,7 @@ setup(void)
 	screen = DefaultScreen(dpy);
 	root = RootWindow(dpy, screen);
 	initfont(font);
-	bh = dc.h = dc.font.height + 2;
+	bh = dc.h = barHeight;
 
 	/* init atoms */
 	wmatom[WMDelete] = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
